### PR TITLE
Type checker

### PR DESCRIPTION
**Pull request by LPeter1997**

### CODE REVIEW

File: src/Draco.Compiler/Internal/Utilities/StringUtils.cs [hunk 1]
```diff
@@ -39,4 +39,22 @@ public static string Unescape(string text)
         }
         return result.ToString();
     }
+
+    /// <summary>
+    /// Converts a 0-based numeric index into an Excel-like column name.
+    /// </summary>
+    /// <param name="index">The index to convert.</param>
+    /// <returns>The Excep column-name of <paramref name="index"/>.</returns>
```
Comment (jakubbinter) [inline src/Draco.Compiler/Internal/Utilities/StringUtils.cs:47]:
```suggestion
    /// <returns>The Excel column-name of <paramref name="index"/>.</returns>
```

File: src/Draco.Compiler/Internal/Codegen/CSharpCodegen.cs [hunk 2]
```diff
@@ -2,369 +2,415 @@
...
+    public override string VisitReferenceExpr(Ast.Expr.Reference node) =>
+        this.AllocateName(node.Symbol);
+
+    private static string? MapUnaryOperator(ISymbol.IUnaryOperator op, string? sub)
+    {
+        if (sub is null) return null;
+
+        if (op == Intrinsics.Operators.Not_Bool) return $"!{sub}";
+
+        throw new NotImplementedException();
```
Comment (jakubbinter) [inline src/Draco.Compiler/Internal/Codegen/CSharpCodegen.cs:332]: Shouldn't we just return the operator?

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/Ast.cs [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+            [property: Ignore(IgnoreFlags.TransformerTransform)] ParseTree? ParseTree,
+            [property: Ignore(IgnoreFlags.TransformerTransform)] object? Value,
+            [property: Ignore(IgnoreFlags.TransformerTransform)] Type Type) : Expr
+        {
+            [Ignore(IgnoreFlags.TransformerAll)]
+            public override Type EvaluationType { get; } = Type;
+        }
+
+        /// <summary>
+        /// An if-expression with an option elSse clause.
```
Comment (jakubbinter) [inline src/Draco.Compiler/Internal/Semantics/AbstractSyntax/Ast.cs:115]:
```suggestion
        /// An if-expression with an optional else clause.
```

Comment (jakubbinter) [reply src/Draco.Compiler/Internal/Semantics/AbstractSyntax/Ast.cs:115]: Why are here and three other doc comments dashes like `if-expression`, but the rest of the doc comments doesn't use these dashes?

**Review by jakubbinter — CHANGES_REQUESTED** [review]

File: src/Draco.Compiler/Internal/Codegen/CSharpCodegen.cs (hunk 2 above)
Comment (LPeter1997) [reply src/Draco.Compiler/Internal/Codegen/CSharpCodegen.cs:332]: In general, the intrinsic can resolve to any "shape" of C# expression currently (look as mod and rem as an example), so to be uniform, no.

File: src/Draco.Compiler/Api/Semantics/SemanticModel.cs [hunk 4]
```diff
@@ -44,11 +43,11 @@ public IEnumerable<Diagnostic> GetAllDiagnostics()
...
-            if (SymbolResolution.ReferencesSymbol(tree))
-            {
-                var sym = SymbolResolution.GetReferencedSymbol(this.db, tree);
-                foreach (var diag in sym.Diagnostics) yield return diag.ToApiDiagnostic(tree);
-            }
+            // Symbol
+            foreach (var diag in SymbolResolution.GetDiagnostics(this.db, tree)) yield return diag.ToApiDiagnostic(tree);
+
+            // Type
+            foreach (var diag in TypeChecker.GetDiagnostics(this.db, tree)) yield return diag.ToApiDiagnostic(tree);
```
[the remark this answers is not in the document]
Comment (LPeter1997) [reply src/Draco.Compiler/Api/Semantics/SemanticModel.cs:50]: I'm not sure mushing this all into a single query improves readability. IIRC I had a bit more logic in this latter loop, so it just formed this way.

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/Ast.cs (hunk 3 above)
Comment (LPeter1997) [reply src/Draco.Compiler/Internal/Semantics/AbstractSyntax/Ast.cs]: Good question, likely hack of consistency.

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+                ParseTree: func,
+                DeclarationSymbol: (ISymbol.IFunction?)SymbolResolution.GetDefinedSymbolOrNull(db, func) ?? throw new InvalidOperationException(),
+                Body: ToAst(db, func.Body)),
+            ParseTree.Decl.Label label => new Ast.Decl.Label(
+                ParseTree: label,
+                LabelSymbol: (ISymbol.ILabel?)SymbolResolution.GetDefinedSymbolOrNull(db, label) ?? throw new InvalidOperationException()),
+            ParseTree.Decl.Variable var => new Ast.Decl.Variable(
+                ParseTree: var,
+                DeclarationSymbol: (ISymbol.IVariable?)SymbolResolution.GetDefinedSymbolOrNull(db, var) ?? throw new InvalidOperationException(),
+                Value: var.Initializer is null ? null : ToAst(db, var.Initializer.Value)),
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs:54]: We could really just have a `SymbolResolution.GetDefinedSymbolExpected<TSymbol>` in the API

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    /// <param name="db">The <see cref="QueryDatabase"/> for the computation.</param>
+    /// <param name="expr">The <see cref="ParseTree"/> to construct the <see cref="Ast"/> from.</param>
+    /// <returns>The <see cref="Ast"/> form of <paramref name="expr"/>.</returns>
+    public static Ast.Expr ToAst(QueryDatabase db, ParseTree.Expr expr) => db.GetOrUpdate(
+        expr,
+        Ast.Expr (expr) => expr switch
+        {
+            ParseTree.Expr.Name name => new Ast.Expr.Reference(
+                ParseTree: name,
+                Symbol: (ISymbol.ITyped)SymbolResolution.GetReferencedSymbol(db, name)),
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs:89]: Same, an expected API should be fine here

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+                ParseTree: call,
+                Called: ToAst(db, call.Called),
+                Args: call.Args.Value.Elements.Select(a => ToAst(db, a.Value)).ToImmutableArray()),
+            ParseTree.Expr.Relational rel => new Ast.Expr.Relational(
+                ParseTree: rel,
+                Left: ToAst(db, rel.Left),
+                Comparisons: rel.Comparisons.Select(c => ToAst(db, c)).ToImmutableArray()),
+            ParseTree.Expr.Unary ury => new Ast.Expr.Unary(
+                ParseTree: ury,
+                Operator: (ISymbol.IUnaryOperator?)SymbolResolution.GetReferencedSymbolOrNull(db, ury) ?? throw new InvalidOperationException(),
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs:113]: Same, expected API

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+                            Expression: ToAst(db, inlineBody.Expression)))),
+                Value: Ast.Expr.Unit.Default),
+            _ => throw new ArgumentOutOfRangeException(nameof(funcBody)),
+        });
+
+    private static Ast.ComparisonElement ToAst(QueryDatabase db, ParseTree.ComparisonElement ce) => db.GetOrUpdate(
+        ce,
+        Ast.ComparisonElement (ce) => new(
+            ParseTree: ce,
+            Operator: (ISymbol.IBinaryOperator?)SymbolResolution.GetReferencedSymbolOrNull(db, ce) ?? throw new InvalidOperationException(),
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs:154]: Same, expected API

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+
+    private static Ast.Expr ToAst(QueryDatabase db, ParseTree.Expr.Binary bin)
+    {
+        var left = ToAst(db, bin.Left);
+        var right = ToAst(db, bin.Right);
+
+        // Binary tree either becomes an assignment or a binary expr
+        if (Syntax.TokenTypeExtensions.IsCompoundAssignmentOperator(bin.Operator.Type))
+        {
+            var @operator = (ISymbol.IBinaryOperator?)SymbolResolution.GetReferencedSymbolOrNull(db, bin) ?? throw new InvalidOperationException();
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs:191]: Same, expected API

File: src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        else if (bin.Operator.Type == TokenType.KeywordOr)
+        {
+            return new Ast.Expr.Or(
+                ParseTree: bin,
+                Left: left,
+                Right: right);
+        }
+        else
+        {
+            var @operator = (ISymbol.IBinaryOperator?)SymbolResolution.GetReferencedSymbolOrNull(db, bin) ?? throw new InvalidOperationException();
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/AbstractSyntax/AstBuilder.cs:222]: Same, expected API

File: src/Draco.Compiler/Internal/Semantics/Symbols/Symbol.cs [hunk 11]
```diff
@@ -0,0 +1,635 @@
...
+                {
+                    scope = scope.Parent;
+                    if (scope is null) return null;
+                }
+
+                var funcDef = scope.Definition;
+                Debug.Assert(funcDef is not null);
+
+                var funcSymbol = SymbolResolution.GetDefinedSymbolOrNull(this.db, funcDef);
+                Debug.Assert(funcSymbol is not null);
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/Symbols/Symbol.cs:329]: Expect API

Comment (LPeter1997) [reply src/Draco.Compiler/Internal/Semantics/Symbols/Symbol.cs:329]: Cast becomes redundant too!

File: src/Draco.Compiler/Internal/Semantics/Types/TypeChecker.cs [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    /// <summary>
+    /// Retrieves the <see cref="ParseTree"/> ancestor that can be used for type inference for <paramref name="tree"/>.
+    /// </summary>
+    /// <param name="db">The <see cref="QueryDatabase"/> for the computation.</param>
+    /// <param name="tree">The <see cref="ParseTree"/> to get the ancestor of.</param>
+    /// <returns>The inferrable ancestor of <paramref name="tree"/>.</returns>
+    private static ParseTree GetInferrableAncestor(QueryDatabase db, ParseTree tree)
+    {
+        var scope = SymbolResolution.GetContainingScopeOrNull(db, tree);
+        Debug.Assert(scope is not null);
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/Types/TypeChecker.cs:178]: Expect API

File: src/Draco.Compiler/Internal/Semantics/Types/TypeChecker.cs [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        // Walk up to the nearest scope that's either global or function
+        while (scope.Kind != ScopeKind.Global && scope.Kind != ScopeKind.Function)
+        {
+            scope = scope.Parent ?? throw new InvalidOperationException();
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/Types/TypeChecker.cs:182]: Maybe just assert instead

File: src/Draco.Compiler/Internal/Semantics/Types/TypeInference.cs [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+        base.VisitCallExpr(node);
+
+        var calledType = TypeChecker.TypeOf(this.db, node.Called);
+        var argsType = node.Args.Value.Elements
+            .Select(a => TypeChecker.TypeOf(this.db, a.Value))
+            .ToImmutableArray();
+        // TODO: We want a "callable" constraint here, but for now we'll go around a bit
+        var clientType = new Type.Function(argsType, new Type.Variable(null));
+        var returnType = ((Type.Function)this.solver.Same(calledType, clientType).Result).Return;
+        this.expressions[node] = returnType;
```
Comment (LPeter1997) [inline src/Draco.Compiler/Internal/Semantics/Types/TypeInference.cs:344]: Missing positional info. Maybe add a `Callable` constraint, even if it just uses `Same` underneath, just so the call-site is nicer?

**Pull request merged**